File: Problem3005/solution.js
/* 
Let n be the number of nodes in the list

Time Complexity: O(n)
Space Complexity: O(1)

A simple and easy two pointers solution.
*/

function ListNode(val, next) {
    this.val = (val===undefined ? 0 : val)
    this.next = (next===undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var middleNode = function(head) {
    let p = head;
    let q = head;
    while(p.next != null){
        if(p.next.next != null){
            p = p.next.next;
        }else{
            p = p.next;
        }
        q = q.next;
    }
    return q;
};

let p = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5, new ListNode(6, null))))))
console.log(middleNode(p))

File: Problem3005/solution.py
""" 
Let n be the number of nodes in the list

Time Complexity: O(n)
Space Complexity: O(1)

A simple and easy two pointers solution.

"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def middleNode(self, head: ListNode) -> ListNode:
        p = head
        q = head
        while p.next is not None:
            if p.next.next is not None:
                p = p.next.next
            else:
                p = p.next
            q = q.next
        return q
    
sol:Solution = Solution()
p = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5, ListNode(6))))))
print(sol.middleNode(p).val)